java/code/src/com/redhat/rhn/manager/setup/ProductSyncManager.java

```diff
@@ -157,6 +166,33 @@ public void addProduct(String productIdent, User user) throws ProductSyncExcepti
         }
     }
 
+    /**
+     * Flush and find a product by identifier without hibernate auto-flushing.
+     * Return the found product or throw an exception in case of error.
+     *
+     * @param productIdent the product identifier
+     * @return Xor with the result on the left side, or with an exception on the right side
+     * @throws ProductSyncException in case of error
+     */
+    private SetupWizardProductDto findProductWithoutAutoFlush(String productIdent) throws ProductSyncException {
+        HibernateFactory.getSession().flush();
+        Xor<SetupWizardProductDto, ProductSyncException> result = HibernateFactory.doWithoutAutoFlushing(() -> {
+            try {
+                return Xor.left(findProductByIdent(productIdent));
+            }
+            catch (ProductSyncException e) {
```

Comment:
`ProductSyncException` seems to be used in this class exclusively, and only in a few places. Outside of this class, the only thing that is consumed out of it is the string message (in the one caller of `addProducts`).

What would you think about turning `ProductSyncException` into a `RuntimeException`, and to return an (`Optional`) `String` message instead of a `Throwable` from `addProducts`, thereby removing the `Xor` contortion here (which is by the way internal to netapi), and the wrapping of `Throwable` into `Optional` (aka "poor-man's Xor") above?

Reply:
> ProductSyncException seems to be used in this class exclusively

That's a great catch which I did not see. Let's make use of it! I don't like the structure of the code (that's why I hid it in the separate method `findProductWithoutAutoflush`).

> (which is by the way internal to netapi)

Yes, I always feel a bit guilty of using it in Uyuni, but I also see other usages in our codebase.